rust_src/src/multibyte.rs

```diff
@@ -88,13 +92,85 @@ impl LispStringRef {
+        width
+    }
+
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions.
+    ///
+    /// With `precision` argument, return the width of longest substring that
+    /// doesn't exceed `precision`, and the number of characters and bytes it
+    /// contains in the returned tuple.
+    pub fn display_width(self, precision: Option<usize>) -> (usize, Option<(usize, usize)>) {
+        // Manually determine if string is unibyte.
+        let len = self.len_chars() as usize;
+        let multibyte = self.len_chars() < self.len_bytes();
+        let distab = unsafe { buffer_display_table() };
+        // Sum width
+        let mut width = 0;
+        // Character index
+        let mut i = 0;
+        // Byte index
+        let mut b = 0;
+
+        while i < len {
+            let (chars, bytes, thiswidth) = unsafe {
+                // If there is a composition, get its id.
+                let (cmp_id, end) = match find_composition(i, None, self.into()) {
+                    Some((_, end, val)) => (
+                        get_composition_id(
+                            i as isize,
+                            b as isize,
+                            (end - i) as isize,
+                            val,
+                            self.into(),
+                        ),
+                        end,
+                    ),
+                    None => (-1, 0),
+                };
+                if cmp_id >= 0 {
+                    // Character is a composition, look it up in the composition table.
+                    let chars = end - i;
+                    let bytes = string_char_to_byte(self.into(), end as isize) - b as isize;
+                    let thiswidth = (*(*composition_table.offset(cmp_id))).width as usize;
+                    (chars, bytes as usize, thiswidth)
+                } else {
+                    // Character is a single codepoint, calculate it if multibyte, otherwise get
+                    // raw byte at b.
+                    let (c, bytes) = if multibyte {
+                        string_char_and_length(self.const_data_ptr().add(b))
+                    } else {
+                        (self.as_slice()[b].into(), 1)
+                    };
+                    (1, bytes, char_width(c as i32, distab) as usize)
+                }
+            };
+
+            // Return if adding character exceeds precision
+            if let Some(precision) = precision {
+                if precision - width < thiswidth {
+                    return (width, Some((i, b)));
+                }
+            }
+
+            width = match width.checked_add(thiswidth) {
+                Some(w) => w,
+                None => string_overflow(),
+            };
+            i += chars;
+            b += bytes;
+        }
+
+        let n = precision.map(|_| (i, b));
```

Comment:
What is this line doing?

Reply:
It's essentially a more compact way of writing
```rust
let n = if precision.is_some() {
    Some((i, b))
} else {
    None
}
```
It seemed a bit verbose to write it that way for something that simple, but I can add a comment like "n is Some((i, b)) if precision is Some" or just rewrite the line to the code above if it's necessary for clarity.

Reply:
@agraven I think part of the confusion is the alias above where you unwrap `precision` but here you are again using the original Option.

If I read this right, `n` is really a list or nil. The current `map` would be fine if there was a comment explaining why precision was involved in making the list. And maybe giving `n` a better name too. The Emacs C has too many shadowy corners and lack luster variable names are part of the problem.